From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nesaak <52047222+Nesaak@users.noreply.github.com>
Date: Fri, 19 Jun 2020 16:01:37 -0400
Subject: [PATCH] Add Disguise System


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index eb4e34a18e462f98c30a61439ec30c5df9e054ca..524369be88dcbef706ff07683408f1491499277d 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -118,6 +118,8 @@ public abstract class EntityLiving extends Entity {
     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
 
+    public int disguisedID = -1; // FishingRPG - Disguised System
+
     @Override
     public float getBukkitYaw() {
         return getHeadRotation();
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
index b5e9e728b7f2d3d0e967b7841f7834abd90696ca..f71de5ff9be58da2298b8edb3faeac87345a13fe 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
@@ -23,7 +23,11 @@ public class PacketPlayOutSpawnEntityLiving implements Packet<PacketListenerPlay
     public PacketPlayOutSpawnEntityLiving(EntityLiving entityliving) {
         this.a = entityliving.getId();
         this.b = entityliving.getUniqueID();
-        this.c = IRegistry.ENTITY_TYPE.a((Object) entityliving.getEntityType());
+        if (entityliving.disguisedID == -1) {
+            this.c = IRegistry.ENTITY_TYPE.a(entityliving.getEntityType());
+        } else {
+            this.c = entityliving.disguisedID;
+        }
         this.d = entityliving.locX();
         this.e = entityliving.locY();
         this.f = entityliving.locZ();
