From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nesaak <52047222+Nesaak@users.noreply.github.com>
Date: Fri, 19 Jun 2020 21:23:47 -0400
Subject: [PATCH] Add Disguise System for Entities


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index e0ab058bf947ea10b37eadf6122292e708bd3809..e8cc2cc39672ffe2891bd1d356ea0a2b024bdaf3 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -202,6 +202,8 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
     // Spigot end
     public boolean shouldBeRemoved; // Paper
 
+    public Packet disguise;
+
     public float getBukkitYaw() {
         return this.yaw;
     }
diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 5cc89c0cf9e9e632212a9653391437cbe9b15031..11560555681e55f22d0e1d31c8f63b7085e911a1 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -277,7 +277,12 @@ public class EntityTrackerEntry {
             // CraftBukkit end
         }
 
-        Packet<?> packet = this.tracker.L();
+        Packet<?> packet;
+        if (this.tracker.disguise == null) {
+            packet = this.tracker.L();
+        } else {
+            packet = this.tracker.disguise;
+        }
 
         this.headYaw = MathHelper.d(this.tracker.getHeadRotation() * 256.0F / 360.0F);
         consumer.accept(packet);
