From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nesaak <52047222+Nesaak@users.noreply.github.com>
Date: Sat, 20 Jun 2020 10:23:27 -0400
Subject: [PATCH] Add a Disguise System for Entities


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index e0ab058bf947ea10b37eadf6122292e708bd3809..5a1f34930dd46cc2e4d35ab3e9b3a1a7585a01ae 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -202,6 +202,18 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
     // Spigot end
     public boolean shouldBeRemoved; // Paper
 
+    // Nesaak start - Disguise System
+    public java.util.function.BiConsumer<EntityPlayer, Entity> disguise;
+    public void refreshEntity() {
+        tracker.trackerEntry.broadcast(new PacketPlayOutEntityDestroy(getId()));
+        for (EntityPlayer trackedPlayer : tracker.trackerEntry.trackedPlayers) {
+            tracker.trackerEntry.a(packet -> {
+                trackedPlayer.playerConnection.sendPacket(packet);
+            }, trackedPlayer);
+        }
+    }
+    // Nesaak end
+
     public float getBukkitYaw() {
         return this.yaw;
     }
diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 5cc89c0cf9e9e632212a9653391437cbe9b15031..62353aa33c8f2ac906cb97a7bf1d57f667defccb 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -277,10 +277,18 @@ public class EntityTrackerEntry {
             // CraftBukkit end
         }
 
-        Packet<?> packet = this.tracker.L();
+        // Nesaak start
+        Packet<?> packet = null;
+        if (this.tracker.disguise == null) {
+            packet = this.tracker.L();
+            consumer.accept(packet);
+        } else {
+            this.tracker.disguise.accept(entityplayer, this.tracker);
+        }
+        // Nesaak end
 
         this.headYaw = MathHelper.d(this.tracker.getHeadRotation() * 256.0F / 360.0F);
-        consumer.accept(packet);
+
         if (!this.tracker.getDataWatcher().d()) {
             consumer.accept(new PacketPlayOutEntityMetadata(this.tracker.getId(), this.tracker.getDataWatcher(), true));
         }
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
index e2f24cae8530e06e5bfe09ebf5a9ea7ee84b16cb..1de4066cba6f959da894b1c0c72967d7bc47f28c 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
@@ -94,6 +94,10 @@ public class PacketPlayOutSpawnEntity implements Packet<PacketListenerPlayOut> {
         packetdataserializer.writeShort(this.h);
     }
 
+    public void setType(EntityTypes type) {
+        this.k = type;
+    }
+
     public void a(PacketListenerPlayOut packetlistenerplayout) {
         packetlistenerplayout.a(this);
     }
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
index 54caee23b6713ac5ebc77f4dd91cbdbc2d6e64b9..fc692315bd734a98a883b854a95fe2be3f41499e 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
@@ -88,6 +88,10 @@ public class PacketPlayOutSpawnEntityLiving implements Packet<PacketListenerPlay
         packetdataserializer.writeShort(this.i);
     }
 
+    public void setType(EntityTypes type) {
+        this.c = IRegistry.ENTITY_TYPE.a(type);
+    }
+
     public void a(PacketListenerPlayOut packetlistenerplayout) {
         packetlistenerplayout.a(this);
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index d16d3fe58e6a45e3023946174d2e137a0670e455..1fdf6ce66f098f678f24e0f9c2145ae92a6b001a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1058,4 +1058,13 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return getHandle().spawnReason;
     }
     // Paper end
+
+    // Nesaak start - disguises
+    public void setDisguise(java.util.function.BiConsumer<EntityPlayer, Entity> disguise) {
+        getHandle().disguise = disguise;
+        if (isValid()) {
+            getHandle().refreshEntity();
+        }
+    }
+    // Nesaak end
 }
